Phase 1: Upgrade JS Checkout to Modular (Part 2/2)

thisisayush opened the pull request:
References #283

Continues #293, To be Merged after

# Part 2 of JS Modular Checkout:
- Adds support for Ajax Based Refresh 

Working Demo:

![screen-recorder-sun-oct-23-2022-13-53-51](https://user-images.githubusercontent.com/22245433/197382257-e347309a-a320-4417-ba5f-d9d3235cd591.gif)

thisisayush commented:
Update: Rebased Branch from the Forked Branch (thisisayush/modular-checkout) [ #293 ], Added Support for Auto Refresh on Order Expire and Fixed Decimal Places

![screen-recorder-tue-nov-01-2022-10-08-59](https://user-images.githubusercontent.com/22245433/199160459-2a8dd451-ef22-4247-9828-78955400b0c7.gif)

DarrenWestwood commented on js/checkout.js:
```diff
@@ -180,6 +200,76 @@ class Blockonomics {
+    }
+
+    refresh_order() {
+        let url = `${this.data.api_url}?get_order=${this.get_url_param('show_order')}&crypto=${this.get_url_param('crypto')}`
```
Please see https://github.com/blockonomics/woocommerce-plugin/pull/276/files#diff-32cdce28ac549728f6a28999f6d874aa4f29e6da1406c28c3134cd8d421a866aL183 for forming the woocommerce url: https://github.com/blockonomics/woocommerce-plugin/blob/3302c63a7faa1346cc1faf6d0671e14a0b86e75e/js/app.js#L183-L200

DarrenWestwood commented on php/Blockonomics.php:
```diff
@@ -631,6 +633,10 @@ public function get_wc_order_received_url($order_id){
     }
 
+    public function get_wc_api_url(){
+        return site_url() . "/wc-api/WC_Gateway_Blockonomics";
```
Please refer to https://github.com/thisisayush/woocommerce-plugin/blob/435854c3589ab2abece71d097f782f343eaf2cd6/php/Blockonomics.php#L361

DarrenWestwood commented on js/checkout.js:
```diff
@@ -180,6 +200,76 @@ class Blockonomics {
+                        this._update_order_params(data)
+                    },
+                    () => {
+                        this._fallback_refresh_order()
```
Let's remove the fallback functions from this PR as this adds overhead

thisisayush replied:
Let's keep the one where there is network error, it helps the plugin fallback to legacy mode in case of Issues like CORS.

thisisayush replied:
Done

DarrenWestwood commented on php/Blockonomics.php:
```diff
@@ -572,6 +572,7 @@ public function get_checkout_context($order, $crypto){
                 // Display Checkout Page
                 $context['order_amount'] = $this->fix_displaying_small_values($order['satoshi']);
                 $context['payment_uri'] = $context['crypto']['uri'] . ":" . $order['address'] . "?amount=" . $context['order_amount'];
+                $context['crypto_rate'] = number_format($order['value']/$context['order_amount'], 2, '.', '');
```
Please add code comment to note we are re-calculating the crypto_rate, which may slightly differ to rate returned by Blockonomics. Please also add reasoning to this PR as to why we are doing this, as opposed to passing this returned value with the $order.

thisisayush replied:
Done

blockonomics commented on js/checkout.js:
```diff
@@ -180,6 +200,70 @@ class Blockonomics {
+        this.reset_progress()
+    }
+
+    _fallback_refresh_order() {
+        location.reload()
+    }
+
+    _satoshi_to_amount(satoshi) {
+        return satoshi/1.0e8
+    }
+
+    _price_per_crypto(crypto, fiat_amount) {
+        return Number(fiat_amount/crypto).toFixed(2) // Show Fixed Decimal Points rather than rounding off
+    }
```
One line functions are a bad practice particularly particularly when the function itself is called from only one place
https://mytwoocents.com/2017/03/25/one-line-methods-can-easily-become-an-antipattern/
In C we have macros for the this same reason

thisisayush replied:
It makes sense, I've updated the PR for the same. I haven't yet replaced `redirect_to_finish_order` method as it was not part of this PR.

thisisayush commented:
@DarrenWestwood 

As discussed, I've separated the `get_order` endpoint which only contains the relevant information required for Updating in JS, it provided us with the flexibility of using the backend for some things we were calculating in the front, such as the crypto rate. and thus I've moved these things to the API to provide better consistency in data.

Also, I meed a crucial part that was updating QR Code and Payment URIs, which I've added in the latest commit.
